src/averageAges.js

```diff
@@ -56,6 +99,41 @@ function calculateWomenAverageAge(people, withChildren) {
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
   // write code here
+  const isOnlySonsSpecified = arguments.length > 1;
+
+  const mothersBirthdayData = people.filter(
+    human => human.sex === 'f').reduce((prev, person) => {
+    return {
+      ...prev,
+      [person.name]: person.born,
+    };
+  }, {});
+
+  const peopleMothersData = people.reduce((prev, person) => {
+    switch (isOnlySonsSpecified) {
+      case false:
+        return person.mother in mothersBirthdayData
+          ? [
+            ...prev,
+            [person.mother, mothersBirthdayData[person.mother], person.born],
+          ]
+
+          : prev;
+
+      case true:
+        return person.mother in mothersBirthdayData && person.sex === 'm'
+          ? [
+            ...prev,
+            [person.mother, mothersBirthdayData[person.mother], person.born],
+          ]
+
+          : prev;
```

Comment:
this code is way too complicated
use filter to first find all mothers, then another filter to find all children (with onlyWithSon check)
then find average age diff